uxarray/grid.py

```diff
@@ -0,0 +1,278 @@
+                raise RuntimeError(msg)
+                exit()
+        except ValueError as e:
+            # check if this is a shp file
+            # we won't use xarray to load that file
+            if file_extension == ".shp":
+                self.mesh_filetype = "shp"
+            else:
+                msg = str(e) + ': {}'.format(self.filepath)
+                print(msg)
+                raise RuntimeError(msg)
+                exit
+
+        # Detect mesh file type, based on attributes of ext_ds:
+        # if ext_ds has coordx or coord - call it exo format
+        # if ext_ds has grid_size - call it SCRIP format
+        # if ext_ds has ? read as shape file populate_scrip_dataformat
+        # TODO: add detection of shpfile etc.
+        if self.ext_ds is None:  # this is None, when xarray was not used to load the input mesh file
+            print(
+                "external file was not loaded by xarray, setting meshfile type to shp"
+            )
+            # TODO: add more checks for detecting shp file
+            self.mesh_filetype = "shp"
+        elif "coordx" in self.ext_ds.data_vars or "coord" in self.ext_ds.data_vars:
+            self.mesh_filetype = "exo"
+        elif "grid_center_lon" in self.ext_ds.data_vars:
+            self.mesh_filetype = "scrip"
+        elif "Mesh2" in self.ext_ds.data_vars:
+            self.mesh_filetype = "ugrid"
+        else:
+            print("mesh file not supported")
+            self.mesh_filetype = None
+
+        return self.mesh_filetype
+
+    # initialize mesh2 DataVariable for uxarray
+    def __init_mesh2__(self):
+        # set default values and initialize Datavariable "Mesh2" for uxarray
+        self.in_ds["Mesh2"] = xr.DataArray(
+            data=0,
+            attrs={
+                "cf_role": "mesh_topology",
+                "long_name": "Topology data of unstructured mesh",
+                "topology_dimension": -1,
+                "node_coordinates": "Mesh2_node_x Mesh2_node_y Mesh2_node_z",
+                "node_dimension": "nMesh2_node",
+                "face_node_connectivity": "Mesh2_face_nodes",
+                "face_dimension": "nMesh2_face"
+            })
+
+    # renames the grid file
+    def saveas_file(self, filename):
+        path = PurePath(self.filepath)
+        new_filepath = path.parent / filename
+        self.filepath = str(new_filepath)
+        self.write_ugrid(self.filepath)
+        print(self.filepath)
+
+    # Calculate the area of all faces.
+    def calculate_total_face_area(self):
+        pass
+
+    # Build the node-face connectivity array.
+    def build_node_face_connectivity(self):
+        pass
+
+    # Build the edge-face connectivity array.
+    def build_edge_face_connectivity(self):
+        pass
+
+    # Build the array of latitude-longitude bounding boxes.
+    def buildlatlon_bounds(self):
+        pass
+
+    # Validate that the grid conforms to the UXGrid standards.
+    def validate(self):
+        pass
```

Comment:
add exceptions to state that the behavior is not yet implemented, instead of quietly `pass`ing.